### ISSUE TITLE
[Question] - Problem with Mvc and Routing

### ISSUE THREAD

**Issue by dazinator** [aspnet/AspNetCore#10441]
I want my application to force a user to go through some initial setup wizard (MVC), before the system is operational. I'll deploy it with a setting, `SetupComplete = false` and once they complete the setup wizard, i'll set that setting to true. Once setup is complete, the rest of the system should become operational - by which I mean the other MVC controllers that I have within the application, can now be successfully routed too - but the Setup wizard (SetupController) will no longer be able to be routed to. 
One last thing to note is that in order to perform the setup, The SetupController will redirect the anonymous user to a Login page provided as an MVC Controller in a nuget package. It will present the user with a login screen, and ask them to enter a code that has been written to a file on the server. It uses cookie authentication to log the user in, then redirect back to the SetupController. This means that my SetupController can / will have some MVC actions that require authentication.

Whilst the system is in `SetupComplete = false` state then:
 - I want to limit which MVC controllers can be routed to. 
    - SetupController
    - LoginController (in referenced nuget package)

- If the routes aren't matched, then I'll add a redirect to the middleware pipeline, redirecting to the default route (SetupController's welcome page)

Given these constraints the way I have tried to model this, is in the middleware pipeline like so:

```csharp
  app.UseAuthentication(); // Needed so that I can allow an admin to login and complete setup.
  app.MapWhen((context) =>
            {
                var opts = context.RequestServices.GetRequiredService<IOptions<PlatformSetupOptions>>();
                return !opts.Value.SetupComplete;
            }, (branch) =>
            {
                // Setup wizard not yet complete
                // Default route should map to SetupController Welcome, no other MVC controllers should be routed.
                branch.UseEndpoints(endpoints =>
                {                   
                   // What do I put here so that only my SetupController can be routed to? Requests for other controllers should 404
                });
                // Route not matched - so do response.Redirect() here..
}

// Back in the default middleware pipeline again..
// Requests should only be allowed to reach this far if PlatformSetupOptions.Value.SetupComplete = true;
app.UseEndpoints(endpoints =>
                {                   
                   // What do I put here so that ALL OTHER (Not SetupController) controllers can be routed to?
                });
```

I thought about using ApplicationParts, but that seems to be setup in ConfigureServices(), so I can't seem to conditionally use one set of parts / mvc assemblies in my branched pipeline, and another set in my default pipeline (correct me if there is way for me to achieve that). 

So my current thinking is that, as all MVC Application parts will be available, in the branch of the middleware pipeline for when setup is not yet complete, I need to figure out how to only allow routing to match a subset of them (i.e SetupController). In the default pipeline (after setup is complete) I haven't found a way for enabling routing to match "all controllers except SetupController" either. The best I am able to think of in that case, is allow it to be routed to, and then put in logic into my SetupController actions that return a 404 if SetupComplete = true - to mimic as if it doesn't exist. Not very nice in my opinion.

How should this feature be modelled in asp.net core? Am I on the right track? Is it possible to model this scenario nicely in asp.net core today?

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @dazinator.
What you're trying to do is very similar to what OrchardCore provides out of the box. You can see how it's done there or even just use OrchardCore for your needs, if that works.

**Comment by dazinator** [comment]
Thanks. I am curious to see how Orchard handles it, so I have raised an issue over there to see if someone can point me to the relevant parts (it's quite a large code base to navigate).

In addition, I did figure out one way to handle it in the end. Basically using MapWhen() with some custom middleware inserted to only allow requests through for certain whitelisted paths (i.e particular controller actions). 
```csharp
   app.UseAuthentication(); // Needed so that I can allow an admin to login and complete setup.
  app.MapWhen((context) =>
            {
                var opts = context.RequestServices.GetRequiredService<IOptions<PlatformSetupOptions>>();
                return !opts.Value.SetupComplete;
            }, (branch) =>
            {
                // Setup not yet complete
                branch.UseSystemSetup(); // my custom middleware to only allow certain-  whitelisted request paths through necessary for system setup functionality. Non approved paths will be redirected to the  /SetupWelcome approved path - handled my MVC part of UseEndPoints() middleware next in pipeline.

                branch.UseEndpoints(endpoints =>
                {                   
                   endpoints.MapDefaultControllerRoute();
                });                
   }
```

**Comment by dazinator** [comment]
I found a nice way to handle this in the end.

I use a global Mvc ResourceFilter to prevent access to controller endpoints when the status of the system is Setup Incomplete. However I do need some Mvc actions to remain operational in this state (the ones involved in completing the setup process) so I found a way to add an attribute above those such that it bypasses the global resource filter allowing them to be reached.
